sway/config/output.c

```diff
@@ -649,8 +621,227 @@ struct output_config *find_output_config(struct sway_output *sway_output) {
+	}
+	if (!cfg->config || cfg->config->adaptive_sync != -1) {
+		sway_log(SWAY_DEBUG, "Trying with adaptive sync disabled for: %s",
+			backend_state->output->name);
+		wlr_output_state_set_adaptive_sync_enabled(state, false);
+		if (search_finish(ctx, output_idx)) {
+			return true;
+		}
+	}
+	// If adaptive sync has not been set, or fallback in case we are on a
+	// backend that cannot disable adaptive sync such as the wayland backend.
+	sway_log(SWAY_DEBUG, "Trying with adaptive sync unset for: %s",
+		backend_state->output->name);
+	state->committed &= ~WLR_OUTPUT_STATE_ADAPTIVE_SYNC_ENABLED;
+	return search_finish(ctx, output_idx);
+}
+
+static bool search_mode(struct search_context *ctx, size_t output_idx) {
+	struct matched_output_config *cfg = &ctx->configs[output_idx];
+	struct wlr_backend_output_state *backend_state = &ctx->states[output_idx];
+	struct wlr_output_state *state = &backend_state->base;
+	struct wlr_output *wlr_output = backend_state->output;
+
+	if (!config_has_auto_mode(cfg->config)) {
+		return search_adaptive_sync(ctx, output_idx);
+	}
+
+	struct wlr_output_mode *preferred_mode = wlr_output_preferred_mode(wlr_output);
+	if (preferred_mode) {
+		sway_log(SWAY_DEBUG, "Trying with preferred mode for: %s", backend_state->output->name);
+		wlr_output_state_set_mode(state, preferred_mode);
+		if (search_adaptive_sync(ctx, output_idx)) {
+			return true;
+		}
+	}
+
+	if (wl_list_empty(&wlr_output->modes)) {
+		state->committed &= ~WLR_OUTPUT_STATE_MODE;
+		return search_adaptive_sync(ctx, output_idx);
+	}
+
+	struct wlr_output_mode *mode;
+	wl_list_for_each(mode, &backend_state->output->modes, link) {
+		if (mode == preferred_mode) {
+			continue;
+		}
+		sway_log(SWAY_DEBUG, "Trying with mode %dx%d@%dmHz for: %s",
+			mode->width, mode->height, mode->refresh, backend_state->output->name);
+		wlr_output_state_set_mode(state, mode);
+		if (search_adaptive_sync(ctx, output_idx)) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static bool search_render_format(struct search_context *ctx, size_t output_idx) {
+	struct matched_output_config *cfg = &ctx->configs[output_idx];
+	struct wlr_backend_output_state *backend_state = &ctx->states[output_idx];
+	struct wlr_output_state *state = &backend_state->base;
+	struct wlr_output *wlr_output = backend_state->output;
+
+	uint32_t fmts[] = {
+		DRM_FORMAT_XRGB2101010,
+		DRM_FORMAT_XBGR2101010,
+		DRM_FORMAT_XRGB8888,
+		DRM_FORMAT_INVALID,
+	};
+	if (render_format_is_bgr(wlr_output->render_format)) {
+		// Start with BGR in the unlikely event that we previously required it.
+		fmts[0] = DRM_FORMAT_XBGR2101010;
+		fmts[1] = DRM_FORMAT_XRGB2101010;
+	}
+
+	const struct wlr_drm_format_set *primary_formats =
+		wlr_output_get_primary_formats(wlr_output, WLR_BUFFER_CAP_DMABUF);
+	bool need_10bit = cfg->config && cfg->config->render_bit_depth == RENDER_BIT_DEPTH_10;
+	for (size_t idx = 0; fmts[idx] != DRM_FORMAT_INVALID; idx++) {
+		if (need_10bit && render_format_is_10bit(fmts[idx])) {
```

Comment:
Shouldn't this condition be `!need_10bit` or am I too tired to review this

Reply:
You're right, I messed up a simplification. There used to be a 10-bit test before this loop to special-case the "maybe we need 10-bit BGR" case, but now just swap the order in the list.